### ISSUE TITLE
Seven separate 'converged' flags are decoupled from whether the solve converged; two of them can stop a solver early

### ISSUE THREAD

**Issue by derrring** [derrring/MFGArchon#1684]
### The finding

A sweep for observables decoupled from the quantities they report on (the class behind #1671,
#1672, #1683) returned 14 candidates. **Seven of them are `converged`.** That is not a
coincidence of the search — `converged` is the field a user reads to decide whether a result is
usable, and in several places it is wired to something else entirely.

Filed as one issue because they share one root structure, not seven unrelated bugs. Each row
below was verified by the same counterfactual: **make the solve worse and show the flag does not
move.**

### Verified

**1. PINN `converged` is a tautology** — `mfg_pinn_solver.py:818`, `hjb_pinn_solver.py:564`,
`fp_pinn_solver.py:656`.

```python
"converged": len(self.training_history.get("total_loss", [])) > 0
```

It is True iff at least one epoch ran. The guard four lines above already raises `RuntimeError`
when that list is empty, so the expression can only ever be True.

| case | `final_loss` | epochs | `converged` |
|---|---|---|---|
| 1 epoch, loss `1e30` | `1.00e+30` | 1 | **True** |
| diverging `1e0 → 1e12` | `1.00e+12` | 5 | **True** |
| genuinely converged | `1.00e-12` | 4 | True |

Positive control: `final_loss` and `epochs_trained`, computed from the same dict on adjacent
lines, do move.

**2. A user abort is reported as convergence** — `coupling/fixed_point_iterator.py:805-807`. A
callback returning `False` sets `converged = True` with `reason="callback_stopped"`.

| | `converged` | iters | `l2distu_rel` |
|---|---|---|---|
| no callback | **False** | 3 | `2.788e-01` |
| callback aborts at iteration 1 | **True** | 1 | **`1.038e+00`** |

The aborted run reports success at an error *larger* than the run that reports failure.

**3. The convergence summary reports "ever converged", not "converged"** —
`utils/convergence/convergence_monitors.py:280`.
`get_convergence_summary()["converged"] = len(converged_iterations) > 0`, collecting every history
entry ever flagged. Appending diverging iterations after a converged one leaves the summary
`True` with a final `u` error of `5e19`.

**4. `convergence_achieved` is never defined anywhere in the package.** Three read sites, zero
assignments — no class attribute, no property, no writer.

**5. Multi-population `converged` never looks at `U`.** `MultiPopulationResult.converged`,
documented as "whether tolerance was reached", is computed from `M_old` only; no `U_old` is ever
captured. A convergence test on one field of a coupled `(U, M)` system.

### The two that can produce wrong numbers, not just wrong beliefs

**6. `FixedPointSolver`'s residual is the damped step** — the comment says
`"Compute residual: ||x_updated - x_current||"` as the residual of `x = G(x)`, but `x_updated` is
the value *after* damping. The residual therefore shrinks with the damping factor independently
of the actual fixed-point error: **turning damping down makes anything "converge."**

**7. The same shape in the coupling convergence metric** —
`utils/performance/monitoring.py:223`, `:492-493`, `stochastic/common_noise_solver.py:438`:
`converged` claims `max(rel, abs) L2 error < tolerance`, but the values compared are the damped
`self.U` / `self.M` assigned at L684-685, not the pre-damping iterates.

These two differ in kind from 1–5. Items 1–5 mislead a reader. Items 6–7 let the solver stop
early and return a non-solution, so they can change the numbers a user publishes.

### What was NOT swept

Nine further candidates from the same sweep have not had their counterfactuals run yet
(monotonicity enforcer, SL Jacobian clamp, backend `fpk_step` clip-then-normalise,
`check_solver_duality` certifying from a class enum, `variance_reduction_factor`,
`compute_hjb_residual(bc_values=...)` ignoring its parameter, and others). They will be added
here or filed separately once verified — recorded now so the list is not lost.

### Suggested shape of a fix

Not proposing one yet, because the seven do not share a remedy — some need the flag deleted, some
need it computed, and 6–7 need a decision about what the residual of a damped fixed-point
iteration should be. What they do share is that **none of them would survive a test that makes
the solve worse and asserts the flag changes**, which is the pinning test any fix owes.

Refs #1671, #1672, #1683

**Comment by derrring** [comment]
**This issue is about code that no longer exists.** Closing candidate.

`alg/neural/` was deleted wholesale — 23,118 lines, 60 files. Per AGENTS.md the deciding measurement was that `boundary_conditions` appeared **zero times** in `alg/neural/`: those solvers were designed before the geometry and boundary-condition layer this library is built on, so they were not BC-aware and had no seam to make them so. The gap was architectural rather than a set of defects to patch.

A local branch `fix/pinn-converged-tautology-1684` carried a fix for this (two commits, including a review round that reset per-run state and corrected a test which encoded the bug). Every file it touches is now absent from `main`:

```
GONE  mfgarchon/alg/neural/pinn_solvers/base_pinn.py
GONE  mfgarchon/alg/neural/pinn_solvers/fp_pinn_solver.py
GONE  mfgarchon/alg/neural/pinn_solvers/hjb_pinn_solver.py
GONE  mfgarchon/alg/neural/pinn_solvers/mfg_pinn_solver.py
GONE  tests/unit/test_alg/test_pinn_converged_1684.py
```

The branch is being pruned. If the neural paradigm returns it returns designed around the geometry/BC layer rather than beside it, and this convergence-reporting defect would have to be re-examined against whatever that is — the old patch would not apply and, more to the point, would be fixing the wrong architecture.

**Comment by derrring** [comment]
Item-by-item status, verified against the tree at `3aa43a21` rather than from memory. This issue
has never carried one, and `updatedAt` reads as "stalled" while five of seven are in fact done --
I misread it that way myself before checking.

| item | state | evidence |
|---|---|---|
| 1. PINN `converged` is a tautology | **resolved by deletion** | the three PINN solvers are gone; zero files reference `mfg_pinn_solver` / `hjb_pinn_solver` |
| 2. user abort reported as convergence | **fixed** | `tests/unit/test_alg/test_callback_abort_is_not_convergence_1684.py` |
| 3. summary reports "ever converged" | **open** | `convergence_monitors.py:280` is still `"converged": len(converged_iterations) > 0` |
| 4. `convergence_achieved` never defined | **fixed** | PR referencing this comment |
| 5. multi-population never looks at `U` | **open** | `multi_population_iterator.py:233` is `np.max(np.abs(M[k] - M_old[k]))`; `U_old` appears nowhere in the file |
| 6. `FixedPointSolver` residual is the damped step | **fixed** | `tests/unit/test_utils/test_fixed_point_residual_is_undamped_1684.py` |
| 7. same shape in the coupling metric | **fixed** | `tests/unit/test_alg/test_coupling_metric_is_the_map_residual_1684.py` |

**Item 4 was worse than "a flag reads False".** The attribute has four read sites and zero writes
anywhere in the package -- measured with a control, since `self.converged =` appears in four files
and the sibling `iterations_run` is genuinely set in `block_iterators.py` and `fictitious_play.py`.
So it is not a rename that missed some call sites; it is the unimplemented half of a pair, sitting
in the same dict literal as the half that works, which is what made it read as implemented.

The failure differed by site and none of them raised:

- the decorator's `getattr(solver, ..., False)` recorded `converged: False` for every tracked run,
  so the report was wrong exactly when the solve succeeded;
- `benchmark_solver`'s `hasattr` was always false, so no verdict was recorded at all -- and that
  loop discarded `solver.solve()`'s return value, which is where the verdict lives;
- `common_noise_solver.py:438` would have raised `AttributeError`, but has never been reached:
  that solver fails earlier in `create_conditional_problem`, now filed as #2191.

**Remaining work is items 3 and 5.** Both are the same shape as the ones already closed -- a
reported status computed from something other than what it names -- and both have a local oracle,
so neither is blocked on anything.

**Comment by derrring** [comment]
**Correction to my item-by-item comment above.** An adversarial review of #2192 refuted its central
claim, and I have verified the refutation against git history rather than accepting it.

I wrote there that `convergence_achieved` "has four read sites and zero writes anywhere in the
package... So it is not a rename that missed some call sites; it is the unimplemented half of a
pair." **The second half of that is false.**

```
da7b8dfc  2025-10-08 22:58  refactor: Unify SolverResult API to use 'converged' (simpler, scipy-style)
53a79ddd  2025-12-06 07:00  chore: Remove deprecated Dt/Dx properties and convergence_achieved parameter
c8edc15e  2025-10-07 13:56  chore: Phase 10.1 - SolverResult return type standardization
```

`git show da7b8dfc^:mfg_pde/utils/solver_result.py` carries `convergence_achieved: bool = False` at
line 48 and `self.convergence_achieved = convergence_achieved` at line 132 — a declared, assigned
dataclass field, which is one of the two forms my regex was supposed to cover. `da7b8dfc` renamed it
to `converged` and left a deprecated property; `53a79ddd` removed the property a month later.

`common_noise_solver.py`'s read was written by `c8edc15e` on 2025-10-07, **one day before the
rename**, against the then-live field. So it is precisely a call site a rename left behind — the
thing I said it wasn't.

**The error was not the regex.** It was that I ran the regex over the CURRENT tree and then stated
the result as a claim about all time. That is the shape this issue is about, committed in the
comment that closes one of its items.

**What the correction does not change.** The two `monitoring.py` sites read the attribute off the
**solver**, and no solver class has ever carried it, so those two were wrong before the rename and
after it. Item 4's fix is unchanged and #2192 stands with a corrected account.

**What it adds.** The 2025-12 removal orphaned three further reads in `examples/` notebooks that
still raise `AttributeError` today; filed as #2195. Nothing in the gate executes those notebooks,
which is why a December removal is still broken in August — the more interesting half of that issue.

The item table above stands otherwise, with items 3 and 5 now also carrying a correction of their
own: see #2193, where the multi-population `err_M` I added measured the **damped** step and so
reintroduced items 6 and 7 in a new location.

**Comment by derrring** [comment]
**7 of 7, verified against `main` at `d1a4c473` rather than from the PR bodies.**

| item | state | evidence on main |
|---|---|---|
| 1. PINN `converged` is a tautology | resolved by deletion | `/usr/bin/grep -rl "mfg_pinn_solver\|hjb_pinn_solver" mfgarchon/` → 0 |
| 2. user abort reported as convergence | fixed | `test_callback_abort_is_not_convergence_1684.py` |
| 3. summary reports "ever converged" | fixed | `convergence_monitors.py:287` is `bool(self.convergence_history[-1]["converged"])` |
| 4. `convergence_achieved` never defined | fixed | zero code read sites; the two remaining matches are prose recording the history |
| 5. multi-population never looks at `U` | fixed | `U_old` captured; `errors` is the per-population max of both fields |
| 6. `FixedPointSolver` residual is the damped step | fixed | `test_fixed_point_residual_is_undamped_1684.py` |
| 7. same shape in the coupling metric | fixed | `fixed_point_iterator.py`, **and** `multi_population_iterator.py:302` is now `M_map[k] - M_old[k]` |

Item 7 needed a second fix. An adversarial review of #2193 found that the item-5 fix had
**reintroduced item 7 at a new site**: `M = (1-r)*M_old + r*M_map`, so the `M - M_old` I measured
was identically `r * (M_map - M_old)`, and turning the damping down bought the convergence verdict.
Measured on the pinning fixture, one sweep: `4.095006e-01 / 2.047503e-01 / 4.095006e-02 /
4.095006e-03` at relaxation `1.0 / 0.5 / 0.1 / 0.01` before, `4.095006e-01` at all four after.
Pinned by an invariance assertion rather than a value.

That is worth recording as the failure mode of this issue rather than a footnote to it: I had the
class in front of me — I wrote the item-by-item table above listing 6 and 7 as fixed — and read it
as a **ledger of what was done** rather than a **constraint on what I was writing next**.

Two other corrections to my earlier comments here, both from review:

- The claim that `convergence_achieved` "was never a rename" is **false**. It was a declared,
  assigned `SolverResult` field until `da7b8dfc` (2025-10-08) renamed it to `converged`, with a
  deprecated property removed by `53a79ddd`. The `common_noise_solver` read predates the rename by
  one day. My regex was correct; running it over the current tree and stating the result as a claim
  about all time was not.
- Three reads that removal orphaned survive in `examples/` notebooks — filed as #2195, along with
  the observation that nothing in the gate executes those notebooks.

Filed while closing this out: **#2191** (`CommonNoiseMFGSolver.solve()` cannot run at all),
**#2195**, **#2196** (`FixedPointIterator` publishes its `np.ones()` preallocation as a measured
error on a sweep-0 divergence — the same "reports a value where nothing was measured" shape as
item 4, at a site this issue never named).

Closed by #2192 and #2193; #1718, the other owner of this structure, closed by #2194.
